src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V63/Eth63ProtocolHandler.cs

```diff
@@ -142,9 +151,40 @@ public override async Task<TxReceipt[][]> GetReceipts(IReadOnlyList<Keccak> bloc
                 return Array.Empty<TxReceipt[]>();
             }
 
-            GetReceiptsMessage msg = new(blockHashes);
-            TxReceipt[][] txReceipts = await SendRequest(msg, token);
-            return txReceipts;
+            int startingReceiptsCountLimit = _getReceiptsCurrentBatchSize;
+            GetReceiptsMessage msg = new(blockHashes.CappedTo(startingReceiptsCountLimit));
+
+            try
+            {
+                Stopwatch sw = Stopwatch.StartNew();
+                TxReceipt[][] txReceipts = await SendRequest(msg, token);
+
+                long elapsed = sw.ElapsedMilliseconds;
+                if (elapsed < GetReceiptsLatencyLowWatermark && txReceipts.Length == startingReceiptsCountLimit)
+                {
+                    _getReceiptsCurrentBatchSize = Math.Min(
+                        (int)Math.Ceiling(startingReceiptsCountLimit * GetReceiptsBatchSizeAdjustmentFactor),
+                        GetReceiptsMaxBatchSize
+                    );
+                }
+                else if (elapsed > GetReceiptsLatencyHighWatermark)
+                {
+                    _getReceiptsCurrentBatchSize = Math.Max(
+                        (int)Math.Floor(startingReceiptsCountLimit / GetReceiptsBatchSizeAdjustmentFactor),
+                        GetReceiptsMinBatchSize
+                    );
+                }
+
+                return txReceipts;
+            }
+            catch (Exception)
+            {
+                _getReceiptsCurrentBatchSize = Math.Max(
+                    (int)Math.Floor(startingReceiptsCountLimit / GetReceiptsBatchSizeAdjustmentFactor),
+                    GetReceiptsMinBatchSize
+                );
+                throw;
+            }
         }
```

Comment:
We have this code very similar in two places, should we make it reusable somehow?

Reply:
Probably